src/raw/requests.rs

```diff
@@ -417,6 +409,7 @@ impl KvRequest for kvrpcpb::RawBatchScanRequest {
     type Result = Vec<KvPair>;
     type RpcResponse = kvrpcpb::RawBatchScanResponse;
     type KeyData = Vec<BoundRange>;
+    type Response = BoxFuture<'static, Result<Self::Result>>;
```

Comment:
Why is this associated type needed?

Reply:
Some requests need a response of streams. This is not a optimal solution, any better is welcome!

https://github.com/tikv/client-rust/blob/40ce5005427f8c627c8e752a37e302bc9eff39b8/src/transaction/requests.rs#L127-L131

Reply:
The reason was in the description of this pull request:

> ### Generic return type for `KvRequest::reduce` and `KvRequest::execute`
> 
> Currently ([f3f3db3](https://github.com/tikv/client-rust/commit/f3f3db349cf4131642baa83c3977490bcc7364c7)), `Transaction::scan` returns a `BoxStream`, which is similar to go client returns an `Iter`. Due to lack of supports of async iterator in Rust, Stream seems to be a great fit for this issue.
> 
> ```rust
>     pub fn scan(&self, _range: impl RangeBounds<Key>) -> BoxStream<Result<KvPair>>
> ```
> 
> However, trait `KvRequest::reduce` returns `BoxFuture<'static, Result<Self::Result>>`, which is not match `BoxStream<Result<KvPair>>`. Therefore, a new associated item `type Response` for `KvRequest` is introduced to make the return type generic. Not quite sure if this is acceptable.

Reply:
Hmm, I think this is a change to the API that needs some thought - it seems a bit weird to return a stream of KvPairs here, because what the underlying code is doing is getting all KvPairs at once and there is only a stream (iirc) because we might get results from multiple regions. I'm not sure what abstraction works best for users.

For now, could you collect the stream into a future using `try_concat` like do for BatchGet and have a future returning an iterator? You'll need to change the return type of `Transaction::Scan` too. I think it is the simplest thing to do for now (especially as the Rust stream API is still in flux). Could you also file an issue to think about this API to find a good long-term solution please?

Reply:
Thanks! 

The previous implementation was just followed what [go client does](https://github.com/tikv/client-go/blob/a3ebdb020c83b86c36377d20d31e15db6168413d/txnkv/store/scan.go#L85-L130). I would change this return type to `Iterator` ASAP.

In addition, I am just thinking about that are there any specification document that any TiKV client implementation must be adhered to? To me, not sure what interface CRUD API should have. As far as I know, there are only common Protobufs defining gRPC requests/responses. That would be great for the whole community if we have some specs. For instance, MongoDB define some [driver specs](https://docs.mongodb.com/ecosystem/drivers/specs/) defining things like [CRUD API](https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#id14) and [service discovery](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst), which unifies behaviors across difference client implementations without unexpected bugs (though still encountered a few corner cases :trollface:).

Reply:
afaik there are no specs. I think it is difficult to have such a spec because different APIs would be idiomatic in different languages. I think it is possible to have multiple clients for one language which present different APIs but use the same protobufs.